### PR TITLE
Combustible items that get too hot start burning

### PR DESCRIPTION
**Pull request by Hinaichigo**
## What this does
Currently we have the concept of autoignition temperature in the code but it doesn't typically get used outside of already-existing fires touching things on their tile. This PR makes it so that certain things can ignite if they get hot enough. As objects themselves don't have a temperature per se, we can use the reagents temperature; if the nutriment in a carrot is 500K then it makes sense to say the carrot is 500K too. So, if the reagents in something reach that something's autoignition temperature, it will start burning. I made all `food/snacks` autoignite at `AUTOIGNITION_ORGANIC` (633.15 K), and gave them a `fire_fuel` of 5 so they don't immediately turn to ash.

This can happen both by the thermal dissipation system and also manual heating by eg. a laser scalpel, So you can laser scalpel a donut to light it on fire.

Also gave lollipop sticks an autoignition temperature of `AUTOIGNITION_PLASTIC`.

## Why it's good
Seems nice expansion to thermal mechanics and makes intuitive sense that if you had a pizza heated to 1000K it would start burning.

## Feedback
For some reason this is much more controversial than I expected. Please explicate any objections below so I can try to assuage them and incorporate feedback.

## Changelog
:cl:
 * rscadd: Combustible items, including food, will ignite if they get too hot (based on their reagent temperature).

### CODE REVIEW

**Comment by boy2mantwicethefam** [comment]
I don't think that's a good idea, being in the middle of a plasmaflood causes all the food in your bag to reach extreme temperatures, having them also turn to ash wouldn't be fun. Doesn't it also cause it to heat up its surroundings via ignition effects?

**Comment by Hinaichigo** [comment]
It's true that this would happen, though being in the bag does add a bit of insulation where it won't happen as fast as say with a donut laying on the ground.
I have a simple idea of how to make that not happen however, if that is preferred.
It compels the question. If you're wearing a hardsuit and a backpack, should we consider the items in your backpack to be insulated by the hardsuit or are you wearing the hardsuit over your backpack?
I can add that tweak to this PR if it's preferred.

> Doesn't it also cause it to heat up its surroundings via ignition effects?

My understanding of the burning object code is that they don't actually have a thermal effect, instead just having a fire overlay, producing smoke and gradually exhausting `fire_fuel`. That's probably something to change.

**Comment by boy2mantwicethefam** [comment]
I think it would be more fun if it didn't affect the user's contents, even if it's not as realistic

**Comment by Hinaichigo** [comment]
I was thinking to check if the wearer is hardsuited or otherwise thermally insulated from the air or not. So if you're wearing only a jumpsuit and walk into a fire with a box of cereal in your pocket it'll start burning. But if you have a hardsuit the icecream in your bag of holding will stay nice and cool. But if you deliberately stick your laser scalpel in there and start blasting the cookie you were saving for later, it'll burn even if you have a hardsuit on.

**Comment by Hinaichigo** [comment]
Okay if everyone hates this idea as-is I'll probably re-do this PR but make it so stuff in your inventory doesn't get affected or something along those lines.

Keep in mind things like drinking glasses and beakers don't have autoignition temperatures so they won't ignite.

But give me some of the reasoning why it's bad so I can understand because honestly I don't get it.

**Comment by Exxion** [comment]
> Doesn't it also cause it to heat up its surroundings via ignition effects?

Yes, for the record.

EDIT: Actually, sorry, no, not passively. It does make existing fires burn hotter, but an atom that's simply burning outside a fire does not produce heat. This is an inconsistency that I was planning to fix if I ever got around to that fire rework, though.

**Comment by Securitymain** [comment]
>Be chef
>Cook something expensive that requires a great deal of labor
>Clown waddles up, holds a lighter to it
>Counter starts burning, all the food burns to cinders, fire alarms go off

**Comment by Hinaichigo** [comment]
It would probably cool off faster than the clown could heat it, or the clown would need to sit there for a while repeatedly holding the lighter to the food. It has to reach the autoignition temperature of 633.15K.
Lol the unexplained (to me) gap between me thinking this is fun and the downvotes is driving me crazy in terms of trying to figure out why. People want to be able to walk around with 2000K cookies and then be able to throw them into space to cool... or something? Is it the aforementioned hardsuit thing where the food you were saving for a snack after you did your high-temperature atmos work gets burned? I can change that if so. Please someone get in here and set me straight.

**Comment by kilozombie** [comment]
>People want to be able to walk around with 2000K cookies and then be able to throw them into space to cool... or something? Is it the aforementioned hardsuit thing where the food you were saving for a snack after you did your high-temperature atmos work gets burned? I can change that if so. Please someone get in here and set me straight.

I think that generally nobody knows why these changes are happening or why they would be interesting. People don't 'want' to walk around with 2000K cookies, worrying about the temperature of food in your backpack sounds esoteric and pointless at best.

I don't understand why we should really care about this PR at all, good or bad. It's so _nothing_ for game balance but will sometimes lead to random specific items (food) being melted by high temperature. Nobody cares. Nobody wants that.

**Comment by kilozombie** [comment]
Some guy outside our house is building a contraption that none of us understand. He's really insistent on building the contraption and it will be able to punch rocks deep into the earth but it's difficult to see the use of such a thing. Also it's loud and we keep having to make sure he doesn't hurt himself or anyone else, and the whole thing is of extremely dubious construction. But he keeps insisting that it would be fuckin baller if you could punch rocks deep into the ground with the contraption

**Comment by Hinaichigo** [comment]
Well for example this makes it so say there's a roaring fire in one part of a room that heats up the air. Then items in other parts of the room can start burning as well and the fire can spread. Or you can light things on fire by heating them up and use them as ignition sources (if the inconsistency Exxion mentioned above is also fixed). You could laser scalpel a donut to get it burning then throw it at a puddle of fuel in sec to try to start a fire in the brig. Flaming shrapnel thrown around by explosions. Reagent containers and food is actually an exploratory step and I've been thinking about expanding it to other items and using different notions of temperature, but we also have to keep performance in mind which is one reason I've kept it limited to those categories. Overall, more potential sandbox options and this actually gives more of a use to the reagent heating/cooling stuff. Seems nice to have.

Why not have a funloving and curious attitude about what could happen?

**Comment by githubuser4141** [comment]
Idea, RIGs should have a smaller, "internal" storage module that can't hold boxes and has less space that has a temperature protection of whatever the suit is rated for. Regular RIGs can receive a temporary "forcefield" that either uses energy and dissipates after a bit, or is a physical material that drapes over you and your equipment to protect it for a short period. Both modules would be optional upgrades though, but atmos/engi hardsuits would start with the internal box.

This probably won't go in but how does it interact with the storage module available for mechs?

**Comment by githubuser4141** [comment]
> > People want to be able to walk around with 2000K cookies and then be able to throw them into space to cool... or something? Is it the aforementioned hardsuit thing where the food you were saving for a snack after you did your high-temperature atmos work gets burned? I can change that if so. Please someone get in here and set me straight.
> 
> I think that generally nobody knows why these changes are happening or why they would be interesting. People don't 'want' to walk around with 2000K cookies, worrying about the temperature of food in your backpack sounds esoteric and pointless at best.
> 
> I don't understand why we should really care about this PR at all, good or bad. It's so _nothing_ for game balance but will sometimes lead to random specific items (food) being melted by high temperature. Nobody cares. Nobody wants that.

Imagine walking out of a plasma fire holding two ball bearings and throwing them at people - causing serious burns. It shouldn't be a problem as long as it behaves logically and as you'd expect. AKa it should have a minimal footprint and not really interfere with regular gameplay - if your backpack items are burning chances are you're long dead, or you're burning backpack items on purpose.

**Comment by west3436** [comment]
I don't support the insistence on reagent temperature realism. However, this PR really is more akin to what I did previously in #33351. Having more things that can catch on fire is always more !!FUN!!, so I support this.

**Comment by pedr69** [comment]
i'm not a fan of every item being burnable because they just vanish and turn into ash when they burn
organic stuff should turn into charcoal and non-organics should turn into slag while retaining their name

**Comment by Eneocho** [comment]
>all food/snacks autoignite at AUTOIGNITION_ORGANIC

I bring your attention to the roburger, made with 90% all robotic head, or Nova bread (and its derivatives) made with a spice so hot it can burn you to death, or roboticist's tears, made with phazon, robot head and some water, or Kebabs/corndog/CORNYDOG/roach on a stick, that have an actual metal rod on the middle, or pickled beets, peanut butter, grape jelly, blether noodle soup, all of which are inside a closed container (glass one for all except the blether, which is contained in metal), or diamond fries, made with diamond carrot (diamond melts at around 900°C), etc.

**Comment by Hinaichigo** [comment]
I take a look at addressing those foods if it seems like this'll get enough support in its current form.
Or maybe remake this with the hardsuit change if that's the main source of contention or another source of contention can be identified and then do that there.

>This probably won't go in but how does it interact with the storage module available for mechs?

Every atom has a thermal insulation value used for these purposes that gets considered. It's multiplicative so a box in a box in a box in a backpack will be more insulated than something just in a backpack. I could also make mechs thermally insulative, and for mobs I could make it take into account their eg. hardsuit insulation.

>i'm not a fan of every item being burnable because they just vanish and turn into ash when they burn
>organic stuff should turn into charcoal and non-organics should turn into slag while retaining their name

Agree about this and we currently have a melting temperature concept distinct from autoignition but it's hardly used at the moment.

**Comment by SonixApache** [comment]
at last, a reason for the 1.000.000.000 fire extinguishers on the station and all the fire simulation shit

**Comment by Volttekka** [comment]
Listen, there's a limit to how much realism should be simulated, and I think we have already exceeded that limit

**Comment by kilozombie** [comment]
The title of this PR is still really misleading! This isn't "combustible items", it's purely reagent containers. Most things in spess aren't reagent containers. This only affects food.

Food isn't even something that is well-known for being flammable. In what universe does food need to be the conduit for a new, interesting temperature system? It also highlights the truly bizarre way that this system is exploiting the fact that food is a reagent container. Food isn't "food", it's an object that contains nutriment and stuff.

If you actually want temperature mechanics that can be used in ingenious, out-of-the-box ways, actually implement a system, stop hijacking an already-janky system and building a broken mess on top of it. Make every single object have a temperature var. Make hot items redder and cold items desaturated. Make hot items hurt to touch with bare hands, and make black gloves very temperature-resistant. Actually make the mechanic instead of fucking with specifically food.

Setting a donut on fire because "it's a reagent container!" makes so much less sense than just, like, making a molotov. Note that in the real world, food is not used as a way to carry fire. We use torches or lighters, we use flammable materials--food is not actually especially flammable. The high water content in a lot of foods makes them terrible candidates for burning.

I don't see anything in the code that prevents, for instance, soup or watermelons from catching on fire. While that's hilarious in the abstract, it's stupid as hell.

**Comment by Hinaichigo** [comment]
@kilozombie (re)read what I said in
https://github.com/vgstation-coders/vgstation13/pull/35274#issuecomment-1782486432
If I was to give every item in the game temperature simulation at once there would probably be performance issues and many bugs due to all of the various ways things can interact in unpredictable ways. Instead we can start small and gradually work our way out and experiment and figure out as we go. I just can't sympathize with the doom and gloom mindset of where this is a horrible thing that will destroy the eating hot pockets during a plasmafire meta as we know it and bring ruin to the codebase or something.

The thing with soup and watermelons is that the liquids in them would evaporate gradually with the application of heat (the watermelon juice would reach about 100 C then the water would boil off, same with soup), and then they would burn. Here the autoignition temperature is far above the water boiling temperature, meaning the water would already be a gas and no longer inside the watermelon by the point the watermelon ignites, but as we don't have that functionality yet, it's a serviceable abstraction. I would like to add evaporation and phase change enthalpy as well but that's another thing to layer onto it. Step by step, bit by bit. If you have a positive outlook this can be a fun thing you know.

I probably am going to do a more granular pass on the food burning temperatures in an attempt to get @Eneocho 's swing vote. I can disable it on soup or jars or other things like that.

**Comment by kilozombie** [comment]
I did read your other post.

>Or you can light things on fire by heating them up and use them as ignition sources (if the inconsistency Exxion mentioned above is also fixed).

This PR only makes reagent containers flammable.

>You could laser scalpel a donut to get it burning then throw it at a puddle of fuel in sec to try to start a fire in the brig.

This isn't very exciting. You could just make a molotov cocktail or an IED or just set the fuel on fire manually. Why a donut? Why a laser scalpel?

>Flaming shrapnel thrown around by explosions.

Shrapnel isn't a reagent container, so this PR doesn't do anything on that subject.

**Comment by kilozombie** [comment]
>I can disable it on soup or jars or other things like that.

We'll be cutting away existing items for days and days and days if you're going to go that route. I am asking you to stop because I think you're wasting your time and also the time of everyone who has to fix the bugs you accidentally create and the edge cases you never test.

**Comment by Hinaichigo** [comment]
Reagent containers are a wider variety of things beyond just food, including syringes, frying pans, drinking glasses and things like that as well. More general than that we we have:
`obj/item/weapon`  - For this there is currently no general way to heat weapons. So even if I implemented them autoigniting, it wouldn't have an impact on much currently.
`obj/item` - Again no way to heat items generally.
`obj` - Too general for now and includes things like decals, effects, forcefields, etc.
Of course I think wood table parts should be able to autoignite, and forks should be able to melt in high temperatures. However, even if I gave them temperature vars, we have no real system for applying heat to these except for `fire_act()`, which already handles ignition for us. I could expand the thermal dissipation system to achieve this, but then we may be simulating thousands more objects every tick, and at a time when the performance is just starting to be recognized as acceptable (see the issue over the stutters), I don't want to rock the boat in this regard until we can do more narrow scale implementation and testing (like this PR), to keep everything running smoothly. People like atomicity and I think it's good to be atomic and incremental. I feel like you are my wife nagging me for a diamond necklace when I bring you home a simple gold ring. Be happy and enjoy what you got for now honeybun.
Even if it doesn't implement it all in one go, it's a nice stepping stone on the way there, don't you think?

I'll push the food-restrictions related thing and ask you and Eneocho to please look it over and tell me if you find it acceptable. And you can tell me what else you wouldn't like to be burnable.
Also there is a nice solution to the kebabs and soups concept; since they drop rods or snack bowls when they are consumed, we can just drop these when they are ashified as well. So the kebab burns up and leaves a rod on the floor. Or a soup burn-evaporates and leaves an empty bowl.

**Comment by kilozombie** [comment]
I don't like this PR and you won't get me to endorse it by integrating little special cases. I think it's silly and I don't want it clogging up the code. Hope that clears it up

edit: I know this comes across as mean, I really don't have anything against you except a perception that you don't test your code very much. But like, obviously this is just me kind of overselling my distaste, it's all a SS13 project that isn't that big a deal. I just think all this stuff is kind of like a forced meme--even workshopping it doesn't make it any more appealing. I don't really see it as inevitable, but it feels like you do? I dunno.

**Comment by Hinaichigo** [comment]
To me it's like, imagine we were doing this 8 years ago. People would be enthusiastic about what we can add, what new mechanics, what new features or simulation we can add to flesh out the game more. They'd be happy to break things occasionally for the sake of trying new things out. But now maybe there's a sentiment that, "no we are content with what we have, we have achieved what needs to be achieved and anything beyond that is just moving away from perfection" That's the idea of stagnation, as once something peaks, it begins to decay. So I don't want to think we've reached the apex or that we've peaked.
In life I never want to become the person who gets into that stagnant mindset of "I like what I like and I don't need nothin' else, get that shit out of my face cuz I ain't interested". It's more fun to always think about what's beyond the next horizon, what else we can do and see and what new things there are to be discovered. I feel like that orientation leads to fun and discovery and allows me to have some passion in life you know what I mean?

**Comment by kilozombie** [comment]
I don't know how else to phrase it: I love new content when it's a good idea done well. If you want me to love this, you would have to come up with a good idea and then do it well. I'm unexcited and distrustful of your ability to code. Even as a non-collab who does not really know how to read code, I have seen the unintended consequences of past PRs, and am utterly unconvinced that anything in this PR--literally anything--is even barely interesting. It is my opinion that you should work on something else.

It's a rabbit hole of realism arguments, edge cases, bad math, and frustrating code audits. Think about how much cyclical arguing you could have about _specific foods_, like soups and watermelons and then potato chip bags, surely those should burn differently, what about reagent containers like soda cans? The cans can't catch on fire and soda isn't flammable so what happens? Fatty foods like french fries are flammable, but the only way the code knows that something is food is its subtype and its contents, so why wouldn't raw fish cutlets be flammable? What about poutine? Ham? If I fill a syringe with nutriment, is it suddenly going to be flammable?

This is all exhausting to argue. If you spend an extraordinary amount of effort you could probably square the circle on all the edge cases, but I don't understand--truly I do not understand--why you would bother.

Hey, here's a bone to throw: make some cool new reagent recipes that require extreme cold or extreme heat. That would be some actual content.

**Comment by Hinaichigo** [comment]
The autoignition temperature is a property of the reagent container itself, not the reagents. So with the soda can thing, no it won't ignite as soda cans don't have an autoignition temperature. If the soda can was made of wood, and you filled it with hot lava, it would burn, which I think makes sense.

In general, food always has reagents it in (except for some bugged situations)
This is why it's convenient to consider the temperature if the food to be the temperature of its reagents, specifically for the case of food. Like if you measure the temperature of a sugarcane plant, compared to measuring the temperature of the sugar in the sugarcane itself, both processes are essentially describing the same thing.

It doesn't have to be an ad hominem thing about me. Yes I have caused bugs, but I've also fixed bugs. Nothing goes in a straight line in life. Like I want you to consider if someone else proposed the exact same thing I'm proposing here, how it would be received. I kind of have the impression that I added too much stuff with this in a short span, sort of like spending "clout" points in that game Democracy if you know it. So people are fatigued by thinking about this concept, indeed about me as you've expressed, and they feel that they don't want to think about it or argue about it anymore and just downvote. To me it's a pity because I have more energy to add stuff but it becomes like pulling teeth and arguing the edge case ramifications and minutia of and facing nebulous resistance over every single thing. There's no need for it really but it's some sort of inertial property of the human brain I guess where people just get tired of thinking about the same thing over and over. Ideally they could just relax and put it out of mind, instead of fatiguing themselves. It's nicer just simply to have fun and be optimistic.

**Comment by kilozombie** [comment]
>Reagent containers are a wider variety of things beyond just food, including syringes, frying pans, drinking glasses and things like that as well.

Well, you're the one who brought up other containers, but apparently this PR doesn't affect them. So again this is just about food.

>It's nicer just simply to have fun and be optimistic.

Optimism is not a senseless, blind adoration of anything new, it's the tendency to presume that good things are coming. I am always optimistic for the codebase as a whole! New features are great to see and I feel like we're getting a steady influx of things to keep things exciting. I am not excited by this. What's the payoff?

**Comment by Hinaichigo** [comment]
This does make it so that reagent containers in general are affected. However, consider that there aren't many non-food reagent containers that _should_ be flammable. Most non-food reagent containers in practice are metal or glass. The main ones that come to mind are maybe paper cups from the water cooler and the wooden mug used to hold certain alcohols.

>payoff

Mainly more dramatic fires. Right now there's not really a nice way for fires to occur without an ignition source.

**Comment by kilozombie** [comment]
>consider that there aren't many non-food reagent containers that should be flammable.

...I've been asking _you_ to consider that

**Comment by Hinaichigo** [comment]
Upon further consideration I have come to the conclusion that this LE BAD, cookies should not burn because that's not something that should happen, nor something we should ever want to or need to see happening.

**Comment by kilozombie** [comment]
I'm sorry to bum you out at all. I'm one loud voice and I'm definitely not the sole authority.

**Comment by kacurasuo** [comment]
> Okay if everyone hates this idea as-is I'll probably re-do this PR but make it so stuff in your inventory doesn't get affected or something along those lines.
> 
> Keep in mind things like drinking glasses and beakers don't have autoignition temperatures so they won't ignite.
> 
> But give me some of the reasoning why it's bad so I can understand because honestly I don't get it.

would be a good addition to cooking on the grill or in a frying pan if it lit on fire in the pan and burnt up with cute fx
what I really want, and which will never ever get coded because it would be horrible to code, is a little plancha that we can cook on with the food sprites shrunk down and requires some attention from the cook or else your food burns

but as for why people are downvoting it? it's because so many of your reagent temp prs have had unintended consequences (adam syndrome)
people don't just hate your prs, but you've gained this reputation for barely testing your prs and with the hot/cold reagent systems you've added they're so poorly tested that they've required multiple passes to reach not-gamebreaking level
so the reason there are so many downvotes, simply, is lack of trust

things like room temperature food instantly killing you or the caramel pills annihilating chemists, in my opinion, demonstrated a clear absence of any testing beyond compiling and strictly "functioning." the glass shard sharpness PR was part of a pattern of adding realism to the game without testing the game consequences until the comments came and demanded that you look up what sharpness did

basically my honest question is did you test this pr? did you make any effort, before opening the PR, to test just how abusable burning up food is? if you did, and you tell us the results of your testing in the PR body, it builds trust

my opinion is that the content of this pr probably wouldn't have unintended consequences, certainly not on the level of some of the past prs, and you should have asked for the trial merge (they updated policy so you can request a trial merge regardless of the emojis)
but like, really actually test your prs

**Comment by pedr69** [comment]
the trialmin fucking broke him
all the bugfixes are closed now

**Comment by ghost** [comment]
That's one way to end this autism.
